pkg/asset/installconfig/azure/validation_test.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/openshift/installer/pkg/ipnet"
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/azure"
+	"github.com/stretchr/testify/assert"
+)
+
+type editFunctions []func(ic *types.InstallConfig)
+
+var (
+	validVirtualNetwork         = "valid-virtual-network"
+	validNetworkResourceGroup   = "valid-network-resource-group"
+	validRegion                 = "centralus"
+	validComputeSubnet          = "valid-compute-subnet"
+	validControlPlaneSubnet     = "valid-controlplane-subnet"
+	validCIDR                   = "10.0.0.0/16"
+	validComputeSubnetCIDR      = "10.0.0.0/24"
+	validControlPlaneSubnetCIDR = "10.0.32.0/24"
+
+	invalidateMachineCIDR = func(ic *types.InstallConfig) {
+		_, newCidr, _ := net.ParseCIDR("192.168.111.0/24")
+		ic.MachineCIDR = &ipnet.IPNet{IPNet: *newCidr}
+	}
+
+	invalidateNetworkResourceGroup = func(ic *types.InstallConfig) {
+		ic.Azure.NetworkResourceGroupName = "invalid-network-resource-group"
+	}
+	invalidateVirtualNetwork     = func(ic *types.InstallConfig) { ic.Azure.VirtualNetwork = "invalid-virtual-network" }
+	invalidateComputeSubnet      = func(ic *types.InstallConfig) { ic.Azure.ComputeSubnet = "invalid-compute-subnet" }
+	invalidateControlPlaneSubnet = func(ic *types.InstallConfig) { ic.Azure.ControlPlaneSubnet = "invalid-controlplane-subnet" }
+	invalidateRegion             = func(ic *types.InstallConfig) { ic.Azure.Region = "eastus" }
+	removeVirtualNetwork         = func(ic *types.InstallConfig) { ic.Azure.VirtualNetwork = "" }
+	removeSubnets                = func(ic *types.InstallConfig) { ic.Azure.ComputeSubnet, ic.Azure.ControlPlaneSubnet = "", "" }
+
+	virtualNetworkAPIResult = &aznetwork.VirtualNetwork{
+		Name: &validVirtualNetwork,
+	}
+	computeSubnetAPIResult = &aznetwork.Subnet{
+		Name: &validComputeSubnet,
+		SubnetPropertiesFormat: &aznetwork.SubnetPropertiesFormat{
+			AddressPrefix: &validComputeSubnetCIDR,
+		},
+	}
+	controlPlaneSubnetAPIResult = &aznetwork.Subnet{
+		Name: &validControlPlaneSubnet,
+		SubnetPropertiesFormat: &aznetwork.SubnetPropertiesFormat{
+			AddressPrefix: &validControlPlaneSubnetCIDR,
+		},
+	}
+)
+
+func validInstallConfig() *types.InstallConfig {
+	return &types.InstallConfig{
+		Networking: &types.Networking{
+			MachineCIDR: ipnet.MustParseCIDR(validCIDR),
+		},
+		Platform: types.Platform{
+			Azure: &azure.Platform{
+				Region:                   validRegion,
+				NetworkResourceGroupName: validNetworkResourceGroup,
+				VirtualNetwork:           validVirtualNetwork,
+				ComputeSubnet:            validComputeSubnet,
+				ControlPlaneSubnet:       validControlPlaneSubnet,
+			},
+		},
+	}
+}
+
+func TestAzureInstallConfigValidation(t *testing.T) {
+	cases := []struct {
+		name           string
+		edits          editFunctions
+		expectedError  bool
+		expectedErrMsg string
+	}{
+		{
+			name:           "Valid virtual network & subnets",
+			edits:          editFunctions{},
+			expectedError:  true,
```

Comment:
I think the better way would be to have only one if these.
```
if test.expectedErrMsg == "" {
    assert.NoErrro(...)
} else {
    assert.Equal(t, err.Error(), test.expectedErrMsg)
}
```

Reply:
I don't see a problem with how it is currently done. Also, it currently uses regular expression matching while your version is explicit. The returned errors are quite long and I don't believe they provide any more value.

Reply:
I don't care if we use asset.Equal or asset.Regex

but we are using 2 fields to convey the same meaning...

https://github.com/openshift/installer/compare/e5ec969b99b16a644619d4bfbce73e8dfb1035df..9ccdf78831373b7ca7628a4ae3dec4733d193cc9#diff-939f10f76443c324a5cf3725abdd8088L94-R94
is exactly the reason where we have a way to create a faulty test case.

When do you think a test-case would want to expect an error, but not care about the message? or expect no error but still expect some err message?

Reply:
I've reworked this into a way that makes more sense.